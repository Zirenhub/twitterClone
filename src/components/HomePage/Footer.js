import {
  HomepageFooter,
  HomepageFooterContainer,
} from '../../styles/HomePageStyles/HomePage.styled';
import homeButton from '../../assets/images/home-svg.svg';
import searchButton from '../../assets/images/search-svg.svg';
import notificationButton from '../../assets/images/bell-svg.svg';
import messagesButton from '../../assets/images/email-svg.svg';

const Footer = () => {
  return (
    <HomepageFooter>
      <HomepageFooterContainer>
        <img src={homeButton} alt="home button"></img>
      </HomepageFooterContainer>
      <HomepageFooterContainer>
        <img src={searchButton} alt="search button"></img>
      </HomepageFooterContainer>
      <HomepageFooterContainer>
        <img src={notificationButton} alt="notification button"></img>
      </HomepageFooterContainer>
      <HomepageFooterContainer>
        <img src={messagesButton} alt="messages button"></img>
      </HomepageFooterContainer>
    </HomepageFooter>
  );
};

export default Footer;
