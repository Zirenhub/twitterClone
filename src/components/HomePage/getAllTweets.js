import { collection, getDocs } from 'firebase/firestore';
import { db } from '../../Firebase';
import getUserInfo from '../ProfilePage/getUserInfo';

const getAllTweets = async () => {
  const returnData = [];

  const querySnapshot = await getDocs(collection(db, 'posts'));

  querySnapshot.forEach(async (doc) => {
    const data = doc.data();
    const arrayFromData = Object.entries(data);
    arrayFromData.forEach((tweet) => {
      const tweetContent = tweet[1];
      // const date = tweetContent.firestoreDate.toDate().toString().slice(0, 21);
      const date = tweetContent.firestoreDate.toDate();
      tweetContent.firestoreDate = date;
    });

    const userInfo = await getUserInfo(doc.id);
    const convertedData = Object.entries(data).map((entry) => {
      return { [entry[0]]: entry[1], userInfo };
    });

    returnData.push(convertedData);
  });

  return returnData;
};

export default getAllTweets;
