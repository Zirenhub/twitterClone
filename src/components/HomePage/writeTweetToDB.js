import { doc, increment, setDoc, updateDoc } from 'firebase/firestore';
import { v4 as uuidv4 } from 'uuid';
import { db } from '../../Firebase';

const writeTweetToDB = async (userID, tweet) => {
  const key = uuidv4();

  try {
    await setDoc(
      doc(db, 'posts', userID),
      {
        [key]: {
          tweet: tweet,
        },
      },
      { merge: true }
    );
    await updateDoc(doc(db, 'users', userID), {
      tweetsNum: increment(1),
    });
  } catch (error) {
    console.log(error);
  }
};

export default writeTweetToDB;
