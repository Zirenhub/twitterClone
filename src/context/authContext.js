import { createContext, useContext, useEffect, useState } from 'react';
import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signOut,
  onAuthStateChanged,
  updateProfile,
} from 'firebase/auth';
import { auth } from '../Firebase';

const UserContext = createContext();

const AuthContextProvider = ({ children }) => {
  const [user, setUser] = useState({});

  const createUser = async (email, password, userName) => {
    const createNewUser = await createUserWithEmailAndPassword(
      auth,
      email,
      password
    );
    await updateProfile(auth.currentUser, {
      displayName: userName,
    });

    return createNewUser;
  };

  const signIn = async (email, password) => {
    return signInWithEmailAndPassword(auth, email, password);
  };

  const logout = () => {
    return signOut(auth);
  };

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (currentUser) => {
      setUser(currentUser);
    });

    return () => {
      unsubscribe();
    };
  }, []);

  return (
    <UserContext.Provider value={{ createUser, user, signIn, logout }}>
      {children}
    </UserContext.Provider>
  );
};

const UserAuth = () => {
  return useContext(UserContext);
};

export { AuthContextProvider, UserAuth, UserContext };
