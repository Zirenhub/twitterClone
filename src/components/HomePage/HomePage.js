import {
  HomepageHeader,
  HomepageMain,
  HomepageSignout,
  HomepageTestPP,
  HomepageTweetsContainer,
  HomepageWriteTweet,
} from '../../styles/HomePageStyles/HomePage.styled';
import { UserAuth } from '../../context/authContext';
import signOut from '../../assets/images/sign-out-svg.svg';
import tweetButton from '../../assets/images/pencil-svg.svg';
import { useNavigate } from 'react-router-dom';
import Footer from './Footer';
import { useEffect, useState } from 'react';
import getAllTweets from './getAllTweets';
import { LoadingStyled } from '../../styles/WelcomePageStyles/Loading.styled';
import {
  ProfileTweetFeedContainer,
  ProfileTweetContainer,
  ProfileTweetContent,
  ProfileWhiteBold,
  ProfileGrayText,
  ProfileWhite,
  ProfileTweetInteractContainer,
  ProfileCommentButton,
  ProfileLikeButton,
  ProfileRetweetButton,
} from '../../styles/ProfilePageStyles/ProfilePage.styled';
import commentIcon from '../../assets/images/comment-svgrepo-com.svg';
import retweetIcon from '../../assets/images/retweet-svgrepo-com.svg';
import likeIcon from '../../assets/images/like-svgrepo-com.svg';

const HomePage = () => {
  const [allTweets, setAllTweets] = useState([]);
  const [mergedData, setMergedData] = useState([]);
  const [loading, setLoading] = useState(false);

  const { logout } = UserAuth();
  const navigate = useNavigate();

  const handleLogout = async () => {
    try {
      await logout();
      navigate('/');
    } catch (error) {
      console.log(error);
    }
  };

  const handleSwitchToProfile = () => {
    navigate('/profile');
  };

  const navigateToTweet = () => {
    navigate('/tweet');
  };

  useEffect(() => {
    const fetchAllTweets = async () => {
      try {
        const res = await getAllTweets();
        if (res) {
          setAllTweets(res);
          setLoading(false);
        }
      } catch (error) {
        console.log(error);
      }
    };

    setLoading(true);
    fetchAllTweets();
  }, []);

  useEffect(() => {
    allTweets.forEach((userTweets) => {
      userTweets.forEach((tweet) => {
        const convertToArray = Object.entries(tweet);
        setMergedData((prevVal) => [...prevVal, convertToArray]);
      });
    });
  }, [allTweets]);

  useEffect(() => {
    const sortedTweets = mergedData.sort((a, b) => {
      return new Date(b[0][1].firestoreDate) - new Date(a[0][1].firestoreDate);
    });
    setMergedData(sortedTweets);
  }, [mergedData]);

  return (
    <>
      {loading ? (
        <LoadingStyled>Loading</LoadingStyled>
      ) : (
        <HomepageMain>
          <HomepageHeader>
            <div style={{ display: 'flex', alignItems: 'center' }}>
              <HomepageTestPP onClick={handleSwitchToProfile}></HomepageTestPP>
              <div style={{ marginLeft: 15 }}>
                <p style={{ fontSize: 15, fontWeight: 'bold' }}>
                  Latest Tweets
                </p>
              </div>
            </div>

            <HomepageSignout onClick={handleLogout}>
              <img src={signOut} alt="sign out button"></img>
            </HomepageSignout>
          </HomepageHeader>
          <HomepageTweetsContainer>
            <HomepageWriteTweet onClick={navigateToTweet}>
              <img src={tweetButton} alt="write tweet button"></img>
            </HomepageWriteTweet>
            <ProfileTweetFeedContainer>
              {mergedData &&
                mergedData.map((tweet) => {
                  return (
                    <ProfileTweetContainer key={tweet[0][0]}>
                      <div style={{ display: 'flex' }}>
                        <HomepageTestPP
                          style={{
                            border: '1px solid red',
                            backgroundColor: '#ffffff',
                            minHeight: 48,
                            minWidth: 48,
                            flexGrow: 1,
                          }}
                        ></HomepageTestPP>
                        <ProfileTweetContent>
                          <div style={{ display: 'flex' }}>
                            <ProfileWhiteBold>
                              {tweet[1][1].userName}
                            </ProfileWhiteBold>
                            <ProfileGrayText style={{ marginLeft: 10 }}>
                              {tweet[0][1].firestoreDate.slice(0, 21)}
                            </ProfileGrayText>
                          </div>

                          <ProfileWhite>{tweet[0][1].tweet}</ProfileWhite>
                        </ProfileTweetContent>
                      </div>

                      <ProfileTweetInteractContainer>
                        <ProfileCommentButton
                          src={commentIcon}
                        ></ProfileCommentButton>
                        <ProfileRetweetButton
                          src={retweetIcon}
                        ></ProfileRetweetButton>
                        <ProfileLikeButton src={likeIcon}></ProfileLikeButton>
                      </ProfileTweetInteractContainer>
                    </ProfileTweetContainer>
                  );
                })}
            </ProfileTweetFeedContainer>
          </HomepageTweetsContainer>
          <Footer></Footer>
        </HomepageMain>
      )}
    </>
  );
};

export default HomePage;
