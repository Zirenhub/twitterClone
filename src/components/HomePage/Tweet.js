import { HomepageTestPP } from '../../styles/HomePageStyles/HomePage.styled';
import { CloseButton } from '../../styles/WelcomePageStyles/SignUp.styled';
import {
  TweetButton,
  TweetContent,
  TweetForm,
  TweetHeader,
  TweetMain,
  TweetPPContainer,
  TweetTextContainer,
} from '../../styles/HomePageStyles/Tweet.styled';
import { useEffect, useState } from 'react';
import { UserAuth } from '../../context/authContext';

const Tweet = () => {
  const [text, setText] = useState('');
  const [canTweet, setCanTweet] = useState(false);

  const { user } = UserAuth();

  const handleText = (e) => {
    setText(e.target.value);
  };

  const handleSubmitTweet = () => {
    console.log(user);
  };

  useEffect(() => {
    text ? setCanTweet(true) : setCanTweet(false);
  }, [text]);

  return (
    <TweetMain>
      <TweetHeader>
        <CloseButton></CloseButton>
        <TweetButton
          bgColor={canTweet ? '#5dbaec' : '#579bbd'}
          color={canTweet ? '#ffffff' : '#9bbecd'}
          disabled={canTweet ? false : true}
          onClick={handleSubmitTweet}
        >
          Tweet
        </TweetButton>
      </TweetHeader>
      <TweetContent>
        <TweetPPContainer>
          <HomepageTestPP></HomepageTestPP>
        </TweetPPContainer>
        <TweetTextContainer>
          <TweetForm>
            <p>What's happening?</p>
            <textarea
              maxLength="250"
              placeholder="Type here!"
              onChange={handleText}
            ></textarea>
          </TweetForm>
        </TweetTextContainer>
      </TweetContent>
    </TweetMain>
  );
};

export default Tweet;
