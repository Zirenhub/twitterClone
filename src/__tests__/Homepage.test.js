import '@testing-library/jest-dom';
import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import { AuthContextProvider } from '../context/authContext';
import HomePage from '../components/HomePage/HomePage';

afterEach(cleanup);

const user = userEvent.setup();

describe('HomePage renders', () => {
  test('Matches snapshot', () => {
    const dom = renderer
      .create(
        <AuthContextProvider>
          <HomePage />
        </AuthContextProvider>
      )
      .toJSON();

    expect(dom).toMatchSnapshot();
  });

  test('Main content renders', async () => {
    render(
      <AuthContextProvider>
        <HomePage />
      </AuthContextProvider>
    );

    expect(screen.getByText(/Hello (.*)$/)).toBeInTheDocument();
  });
});
