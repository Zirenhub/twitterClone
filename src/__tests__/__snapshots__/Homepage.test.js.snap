// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage renders Matches snapshot 1`] = `
<h1>
  Hello 
</h1>
`;
